### PR TITLE
gh-67224: Make linecache imports relative to improve startup speed

### PR DESCRIPTION
**Pull request by pablogsal**
* Issue: gh-67224

### CODE REVIEW

File: Lib/linecache.py [hunk 1]
```diff
@@ -48,6 +45,7 @@ def getlines(filename, module_globals=None):
 
 
 def checkcache(filename=None):
+    import os
```
Comment (vstinner) [inline Lib/linecache.py]: Please move imports after docstrings.

**Comment by pablogsal** [comment]
Oh man, seems this is complicated because making the import lazy actually fails at shutdown because if the first time this is called its when the interpreter is shutting down, sys.meta_path is None :(

**Comment by pablogsal** [comment]
Well, I think this should do the trick

File: Lib/linecache.py [hunk 2]
```diff
@@ -67,15 +64,22 @@ def checkcache(filename=None):
         if mtime is None:
             continue   # no-op for files loaded via a __loader__
         try:
+            # This import can fail if the interpreter is shutting down
+            import os
             stat = os.stat(fullname)
         except OSError:
             cache.pop(filename, None)
             continue
+        except ImportError:
```
Comment (vstinner) [inline Lib/linecache.py]: Please add a comment explaining how these imports can fail.

File: Lib/linecache.py [hunk 3]
```diff
@@ -67,15 +64,22 @@ def checkcache(filename=None):
...
+            return
         if size != stat.st_size or mtime != stat.st_mtime:
             cache.pop(filename, None)
 
 
 def updatecache(filename, module_globals=None):
+    import os
```
Comment (vstinner) [inline Lib/linecache.py:84]: It may be worth it to add a comment explaining why imports are not top level.

File: Lib/linecache.py (hunk 2 above)
Comment (pablogsal) [reply Lib/linecache.py]: There is already a comment some lines before (before the import itself).

Comment (vstinner) [reply Lib/linecache.py]: Oh right, I reviewed your PR on a phone and I missed it.

File: Lib/linecache.py [hunk 4]
```diff
@@ -67,15 +64,25 @@ def checkcache(filename=None):
         if mtime is None:
             continue   # no-op for files loaded via a __loader__
         try:
+            # This import can fail if the interpreter is shutting down
+            import os
```
Comment (vstinner) [inline Lib/linecache.py:68]: Maybe move the import/except ImportError out of the try/except OSError?

**Review by vstinner — APPROVED** [review]
LGTM. Just a suggestion to reorganize checkcache() import/except ImportError.

File: Lib/linecache.py [hunk 5]
```diff
@@ -67,15 +64,25 @@ def checkcache(filename=None):
...
             stat = os.stat(fullname)
         except OSError:
             cache.pop(filename, None)
             continue
+        except ImportError:
+            return
```
Comment (pablogsal) [inline Lib/linecache.py]:
```suggestion
        try:
            # This import can fail if the interpreter is shutting down
            import os
        except ImportError:
            return
        try:
            stat = os.stat(fullname)
        except OSError:
            cache.pop(filename, None)
            continue
```

File: Lib/linecache.py [hunk 6]
```diff
@@ -76,6 +78,12 @@ def checkcache(filename=None):
 
 
 def updatecache(filename, module_globals=None):
+    # These imports are not at top level because linecache is in the critical
+    # path of the interpreter startup and importing os and sys take a lot of time
+    # and slow down the startup sequence.
+    import os
+    import sys
+
     """Update a cache entry and return its list of lines.
```
Comment (JelleZijlstra) [inline Lib/linecache.py:87]: We lost the docstring because you put the imports before it. Filed #117948 to fix this.

**Pull request merged**